### PR TITLE
Collection links (#91)

### PR DESCRIPTION
**Pull request by retorquere**
This implements #91. It builds on #102 -- if so desired, I can rebase onto #102 after that has been merged.

### CODE REVIEW

**Comment by retorquere** [comment]
https://0x0.st/zOaj.xpi

**Comment by bjohas** [comment]
Thank you - looks great!

A couple of requests:
- Could it also generate the “https://zotero.org/…” links from collections ("Copy Zotero URI")?
- I believe the format e.g. zotero://select/groups/5/collections/JBQWA3FR is an 'old' link, c.f. here: https://github.com/willsALMANJ/Zutilo/issues/97. It might be that once that issue makes it into the release, it will be automatically fixed here. Or maybe not? Would be good to check!

**Comment by bjohas** [comment]
This is not directly related, but in case it's of interest: https://forums.zotero.org/discussion/78550/getting-zotero-to-work-under-ubuntu-linux

**Comment by retorquere** [comment]
It isn't related, but if you installed from my debs, you can file an issue there. If you installed from the tarball, they don't work out of the box.

**Comment by retorquere** [comment]
From that forum post, I don't understand how this would be an old url? What's the difference between the two? It wouldn't be automatically fixed, no.

**Comment by bjohas** [comment]
Ah yes - sorry, not very clear at all. I meant that in zotero://select/groups/5/collections/3ESW8QB9, the '5' refers to a property of my zotero client (being the 5th library perhaps). This was the case with some outdated zotero:// links as well (I think).

Let me start again: For example, in one of my libraries, I have these URIs:
- https://www.zotero.org/groups/2339240/edtechhub/items/collectionKey/3ESW8QB9
- zotero://select/groups/5/collections/3ESW8QB9

I was expecting zotero://select/groups/2339240/collections/3ESW8QB9 for the 2nd.

**Comment by retorquere** [comment]
https://0x0.st/zOSl.xpi

**Comment by retorquere** [comment]
>     * Could it also generate the “[https://zotero.org/…”](https://zotero.org/%E2%80%A6%E2%80%9D) links from collections ("Copy Zotero URI")?

Also as in put them on the clipboard together? Or as in add a new menu option?
 
>     * I believe the format e.g. zotero://select/groups/5/collections/JBQWA3FR is an 'old' link, c.f. here: #97 (and https://forums.zotero.org/discussion/78312/zotero-uri-vs-select-item). It might be that once that issue makes it into the release, it will be automatically fixed here. Or maybe not? Would be good to check!

I think that's fixed in zOSl.xpi

**Comment by retorquere** [comment]
Oh I see what you mean with the URI. https://0x0.st/zOSR.xpi

**Comment by bjohas** [comment]
Perfect - that's fixed it! And I installed from your repo, so the zotero:// links work too!

Will you be able to add https://www.zotero.org/groups/2339240/edtechhub/items/collectionKey/3ESW8QB9 - style links as well? (No rush - just checking it's on the radar)

**Comment by retorquere** [comment]
Doesn't zOSR already add that?

**Comment by bjohas** [comment]
AH YES! It does - as long as I remember to turn it on the in the preferences!! 

THANK YOU!!!!

**Comment by retorquere** [comment]
The PR has been rebased

**Comment by retorquere** [comment]
Shouldn't make any difference, but to make sure: https://0x0.st/zVTC.xpi

**Comment by bjohas** [comment]
@willsALMANJ  - just a tiny friendly and most courteous prod on this PR :)
Thanks!!

**Comment by bjohas** [comment]
@retorquere I think the collection links may be wrong.

When I copy the URI to a collection
https://www.zotero.org/groups/2292090/activating_edtech_1_resources/collections/XGMUY62W?
However, that link just goes to the main library.

When I actually click on the lhe link, I get
https://www.zotero.org/groups/2292090/activating_edtech_1_resources/items/collectionKey/XGMUY62W
which shows the items. From then on, when I hover, the correct links come up. It may be a slight inconsistency in the web library system (which is changing anyway).

I realise that this is my mistake, as I suggested the wrong link - sorry. Can you adjust the code to use the alternative link instead?

**Comment by retorquere** [comment]
That's pretty strange because I just request the collection URI from Zotero using `Zotero.URI.getCollectionURI`

**Comment by willsALMANJ** [comment]
@bjohas thanks -- I haven't forgotten. I am trying to get to this as soon as I can.

**Comment by willsALMANJ** [comment]
I reviewed this and it looked good to me (thanks for incidentally improving the code as you add things to it).

One thing I noticed is that I get this message in the error console when clicking on an item in the collection context menu:

```
Timestamp: 08/23/2019 12:31:23 AM
Error: TypeError: o is undefined
Source File: chrome://zotero/content/zoteroPane.js
Line: 2372
```

referencing [this line](https://github.com/zotero/zotero/blob/bb594296643c50bc6ebc80272a9ba7cf2469d5a3/chrome/content/zotero/zoteroPane.js#L2372). It seems that Zotero is trying to capture the click event and handle it in JS instead of using the built-in XUL system for tying commands to menu items. The copy functions still work despite the error. I am not sure there is an easy to prevent Zotero from trying to handle the click, so we might just have to leave it the way it is. What do you think?

Besides that, we just need to resolve the question about the URI format, and I can copy the strings to the other locales and merge. I haven't read the clean up PR yet, but I don't anticipate that being a problem to merge after I get a chance to read it. We should add something to the documentation for these functions as well (can be done separately from this PR).

**Comment by retorquere** [comment]
I've asked about `Zotero.URI.getCollectionURI` on the Zotero dev forums. When I get more info, I'll amend the copy action.

It's going to be hard to get rid of the `o is undefined` error, as that function searches a private array. I could monkey-patch [onCollectionContextMenuSelect](https://github.com/zotero/zotero/blob/bb594296643c50bc6ebc80272a9ba7cf2469d5a3/chrome/content/zotero/zoteroPane.js#L2370) to only pass on events when they don't match IDs known to Zutilo, but monkey-patching has it's own risks, and since `onCollectionContextMenuSelect` is only called [here](https://github.com/zotero/zotero/blob/8f00fa2603300e6e1a00893afc6bed2a55ba372c/chrome/content/zotero/zoteroPane.xul#L228), the error disappears harmlessly into the void.

**Comment by retorquere** [comment]
https://groups.google.com/forum/#!msg/zotero-dev/wryVEd_mJNM/pL5_Do48AgAJ

**Comment by retorquere** [comment]
So it looks like it's something that's going to be fixed at their end.

**Comment by retorquere** [comment]
https://0x0.st/zW3g.xpi disables the "copy link" option if the user isn't synced.

The cleanup PR can wait until we've settled this one. I'll rebase to master after this is merged and then re-run the linter.

File: addon/chrome/content/zutilo/zoteroOverlay.js [hunk 1]
```diff
@@ -1195,12 +1217,11 @@ ZutiloChrome.zoteroOverlay = {
...
-    zoteroItemmenuItem: function(functionName, IDPrefix, doc) {
+    zoteroMenuItem: function(menuName, functionName, IDPrefix, doc) {
         var menuFunc = doc.createElement('menuitem');
         menuFunc.setAttribute('id', IDPrefix + functionName);
         menuFunc.setAttribute('label',
-            Zutilo._bundle.GetStringFromName('zutilo.itemmenu.' +
-                                             functionName));
+            Zutilo._bundle.GetStringFromName(`zutilo.${menuName}menu.${functionName}`));
         menuFunc.addEventListener('command',
             function() {
```
Comment (willsALMANJ) [inline addon/chrome/content/zutilo/zoteroOverlay.js]: I think if we made this listener callback here take the `event` argument here and call `event.stopPropagation()` first that the event would not get to the `oncommand` `onCollectionContextMenuSelect` callback on the parent menu and then we wouldn't get the error. I haven't thought enough about if there is any downside to this.

**Comment by willsALMANJ** [comment]
Okay, so then we are good to merge this other than the benign error. I posted a comment in-line about a way to avoid the error. I am curious what you think.

File: addon/chrome/content/zutilo/zoteroOverlay.js (hunk 1 above)
Comment (retorquere) [reply addon/chrome/content/zutilo/zoteroOverlay.js]: That should work

**Comment by retorquere** [comment]
https://0x0.st/zWl2.xpi

**Comment by bjohas** [comment]
Thank you both, that's amazing!

**Pull request merged**